README.md
# React_Generic Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/React_Generic.svg?)](https://www.ardu-badge.com/React_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/React_Generic.svg)](https://github.com/khoih-prog/React_Generic/releases)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/React_Generic.svg)](http://github.com/khoih-prog/React_Generic/issues)
![C++](https://img.shields.io/badge/language-C++-blue.svg)
[![license: MIT](https://img.shields.io/badge/license-MIT-brightgreen.svg)](https://opensource.org/licenses/MIT)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents

* [Why do we need this React_Generic library](#why-do-we-need-this-React_Generic-library)
  * [Features](#features)
  * [Currently supported Boards](#currently-supported-boards)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [HOWTO Use](#howto-use)
  * [Blink](#Blink)
  * [Why Bother?](#Why-Bother)
  * [Advanced callback support](#Advanced-callback-support)
* [API](#API)
  * [Namespace use](#Namespace-use)
  * [Event Registration Functions](#Event-Registration-Functions)
  * [Management functions](#Management-functions)
* [Examples](#examples)
  * [Original Examples](#original-examples)
    * [ 1. minimal](examples/minimal)
    * [ 2. torture_test](examples/torture_test)
  * [multiFileProject Examples](#multiFileProject-examples)
    * [ 1. multiFileProject](examples/multiFileProject) 
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [License](#license)
* [Copyright](#copyright) 
  
---
---

### Why do we need this [React_Generic library](https://github.com/khoih-prog/React_Generic)

#### Features

By 

- [Andrew Carter](https://github.com/AndrewCarterUK)
- [Matti Airas](https://github.com/mairas)
- [Khoi Hoang](https://github.com/khoih-prog)

An asynchronous programming library for the generic microcontrollers using the Arduino framework. Currently supporting ESP32, ESP8266, SAMD21/SAMD51, RP2040, STM32, nRF52, Teensy 4.x

The library is at the core of the [SensESP](https://github.com/SignalK/SensESP) project but is completely generic and can be used for standalone projects without issues.

This library gets much of its inspiration (and some code) from [`Reactduino`](https://github.com/Reactduino/Reactduino). `React_Generic`, however, has been internally re-implemented for maintainability and readability, and has significantly better performance when there are lots of defined reactions. It also supports arbitrary callables as callbacks, allowing parametric creation of callback functions.

#### Currently supported Boards

This [**React_Generic** library](https://github.com/khoih-prog/React_Generic) currently supports these following boards:

 1. **ESP32**
 
 2. **ESP8266**
 
 3. **WT32_ETH01 (ESP32 + LAN8720A)**
 
 4. **STM32F/L/H/G/WB/MP1 boards with built-in Ethernet LAN8742A** such as :

  - **Nucleo-144 (F429ZI, F767ZI)**
  - **Discovery (STM32F746G-DISCOVERY)**
  - **All STM32 boards (STM32F/L/H/G/WB/MP1) with 32K+ Flash, with Built-in Ethernet**
  
 5. **STM32F4/F7 boards using Ethernet LAN8720** such as :

  - **Nucleo-144 (F429ZI, NUCLEO_F746NG, NUCLEO_F746ZG, NUCLEO_F756ZG)**
  - **Discovery (DISCO_F746NG)**
  - **STM32F4 boards (BLACK_F407VE, BLACK_F407VG, BLACK_F407ZE, BLACK_F407ZG, BLACK_F407VE_Mini, DIYMORE_F407VGT, FK407M1)** 
  
  `Note: Must use with STM32 core v2.2.0-`
  
 6. **SAMD21/SAMD51**
 
 7. **nRF52 boards**, such as **AdaFruit Feather nRF52832, nRF52840 Express, BlueFruit Sense, Itsy-Bitsy nRF52840 Express, Metro nRF52840 Express, NINA_B302_ublox, NINA_B112_ublox, etc.**
 
 8. **Teensy 4.1 using QNEthernet Library**
 
 9. **RASPBERRY_PI_PICO** using [**arduino-pico core v2.5.2+**](https://github.com/earlephilhower/arduino-pico)
  

---
---

## Prerequisites

 1. [`Arduino IDE 1.8.19+` for Arduino](https://github.com/arduino/Arduino). [![GitHub release](https://img.shields.io/github/release/arduino/Arduino.svg)](https://github.com/arduino/Arduino/releases/latest)
 2. [`ESP32 Core 2.0.45`](https://github.com/espressif/arduino-esp32) for ESP32-based boards. [![Latest release](https://img.shields.io/github/release/espressif/arduino-esp32.svg)](https://github.com/espressif/arduino-esp32/releases/latest/)
 3. [`ESP8266 Core 3.0.2+`](https://github.com/esp8266/Arduino) for ESP8266-based boards. [![Latest release](https://img.shields.io/github/release/esp8266/Arduino.svg)](https://github.com/esp8266/Arduino/releases/latest/).
 4. [`Adafruit nRF52 v1.3.0+`](https://github.com/adafruit/Adafruit_nRF52_Arduino) for nRF52 boards such as Adafruit NRF52840_FEATHER, NRF52832_FEATHER, NRF52840_FEATHER_SENSE, NRF52840_ITSYBITSY, NRF52840_CIRCUITPLAY, NRF52840_CLUE, NRF52840_METRO, NRF52840_PCA10056, PARTICLE_XENON, **NINA_B302_ublox**, etc. [![GitHub release](https://img.shields.io/github/release/adafruit/Adafruit_nRF52_Arduino.svg)](https://github.com/adafruit/Adafruit_nRF52_Arduino/releases/latest)
 5. [`Teensy core v1.57+`](https://www.pjrc.com/teensy/td_download.html) for Teensy (4.1, 4.0, 3.6, 3.5, 3,2, 3.1, 3.0) boards.
 6. [`ArduinoCore-mbed mbed_rp2040, mbed_nano, mbed_portenta core 3.4.1+`](https://github.com/arduino/ArduinoCore-mbed) for Arduino (Use Arduino Board Manager) **Portenta_H7, RP2040-based boards, such as Nano_RP2040_Connect, RASPBERRY_PI_PICO**. [![GitHub release](https://img.shields.io/github/release/arduino/ArduinoCore-mbed.svg)](https://github.com/arduino/ArduinoCore-mbed/releases/latest)
 7. [`Earle Philhower's arduino-pico core v2.6.3+`](https://github.com/earlephilhower/arduino-pico) for RP2040-based boards such as **RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, etc. [![GitHub release](https://img.shields.io/github/release/earlephilhower/arduino-pico.svg)](https://github.com/earlephilhower/arduino-pico/releases/latest)
 8. [`Arduino SAM DUE core v1.6.12+`](https://github.com/arduino/ArduinoCore-sam) for SAM DUE ARM Cortex-M3 boards.
 9. [`Arduino SAMD core 1.8.13+`](https://github.com/arduino/ArduinoCore-samd) for SAMD ARM Cortex-M0+ boards. [![GitHub release](https://img.shields.io/github/release/arduino/ArduinoCore-samd.svg)](https://github.com/arduino/ArduinoCore-samd/releases/latest)
10. [`Adafruit SAMD core 1.7.11+`](https://github.com/adafruit/ArduinoCore-samd) for SAMD ARM Cortex-M0+ and M4 boards (Nano 33 IoT, etc.). [![GitHub release](https://img.shields.io/github/release/adafruit/ArduinoCore-samd.svg)](https://github.com/adafruit/ArduinoCore-samd/releases/latest)
11. [`Seeeduino SAMD core 1.8.3+`](https://github.com/Seeed-Studio/ArduinoCore-samd) for SAMD21/SAMD51 boards (XIAO M0, Wio Terminal, etc.). [![Latest release](https://img.shields.io/github/release/Seeed-Studio/ArduinoCore-samd.svg)](https://github.com/Seeed-Studio/ArduinoCore-samd/releases/latest/)

---
---

## Installation

### Use Arduino Library Manager

The best and easiest way is to use `Arduino Library Manager`. Search for `React_Generic`, then select / install the latest version.
You can also use this link [![arduino-library-badge](https://www.ardu-badge.com/badge/React_Generic.svg?)](https://www.ardu-badge.com/React_Generic) for more detailed instructions.

### Manual Install

1. Navigate to [React_Generic](https://github.com/khoih-prog/React_Generic) page.
2. Download the latest release `React_Generic-main.zip`.
3. Extract the zip file to `React_Generic-main` directory 
4. Copy the whole `React_Generic-main` folder to Arduino libraries' directory such as `~/Arduino/libraries/`.

### VS Code & PlatformIO:

1. Install [VS Code](https://code.visualstudio.com/)
2. Install [PlatformIO](https://platformio.org/platformio-ide)
3. Install [**React_Generic** library](https://registry.platformio.org/libraries/khoih-prog/React_Generic) by using [Library Manager](https://registry.platformio.org/libraries/khoih-prog/React_Generic/installation). Search for React_Generic in [Platform.io Author's Libraries](https://platformio.org/lib/search?query=author:%22Khoi%20Hoang%22)
4. Use included [platformio.ini](platformio/platformio.ini) file from examples to ensure that all dependent libraries will installed automatically. Please visit documentation for the other options and examples at [Project Configuration File](https://docs.platformio.org/page/projectconf.html)

---
---

## HOWTO Use

### Blink

If you have worked with the Arduino framework before, it is likely that you will have come across the [blink sketch](https://www.arduino.cc/en/tutorial/blink). This is a simple program that flashes an LED every second, and it looks something like this:

```cpp
#include <Arduino.h>

void setup() 
{
  pinMode(LED_BUILTIN, OUTPUT);
}

void loop() 
{
  digitalWrite(LED_BUILTIN, HIGH);
  delay(1000);
  digitalWrite(LED_BUILTIN, LOW);
  delay(1000);
}
```

Using React_Generic, the sketch can be rewritten to the following:

```cpp
#include <React_Generic.h>

using namespace react_generic;

React_Generic app;

setup() 
{
  pinMode(LED_BUILTIN, OUTPUT);

  app.onRepeat(1000, [] () 
  {
      static bool state = false;
      digitalWrite(LED_BUILTIN, state = !state);
  });
}

void loop() 
{
  app.tick();
}
```

Instead of directly defining the program logic in the `loop()` function, _reactions_ are defined in the `setup()` function. A reaction is a function that is executed when a certain event happens. In this case, the event is that the function should repeat every second, as defined by the `onRepeat()` method call. The second argument to the `onRepeat()` method is a [lambda function](http://en.cppreference.com/w/cpp/language/lambda) that is executed every time the reaction is triggered. If the syntax feels weird, you can also use regular named functions instead of lambdas.

The `app.tick()` call in the `loop()` function is the main loop of the program. It is responsible for calling the reactions that have been defined. You can also add additional code to the `loop()` function, any delays or other long-executing code should be avoided.

---

### Why Bother?

Charlie wants to make a simple program which echoes data on the `Serial` port. Their Arduino sketch will looks like this:

```cpp
#include <Arduino.h>

void setup()
{
    Serial.begin(9600);
}

void loop()
{
    if (Serial.available() > 0) 
    {
        Serial.write(Serial.read());
    }

    yield();
}
```

This works, but Charlie decides that they would like to blink the built-in LED every time it processes data. Now, their sketch looks like this:

```cpp
#include <Arduino.h>

void setup()
{
    Serial.begin(9600);
    pinMode(LED_BUILTIN, OUTPUT);
}

void loop()
{
    if (Serial.available() > 0) 
    {
        Serial.write(Serial.read());

        digitalWrite(LED_BUILTIN, HIGH);
        delay(20);
        digitalWrite(LED_BUILTIN, LOW);
    }

    yield();
}
```

The problem with this sketch is that whilst the LED is blinking, Charlie's program is not relaying data from the Serial port. The longer Charlie blinks the LED for, the slower the rate of transfer.

To solve this problem, Charlie refactors their code to look something like this:

```cpp
#include <Arduino.h>

uint32_t start;
bool blink = false;

void setup()
{
    Serial.begin(9600);
    pinMode(LED_BUILTIN, OUTPUT);
}

void loop()
{
    if (Serial.available() > 0) 
    {
        Serial.write(Serial.read());

        blink = true;
        start = millis();
        digitalWrite(LED_BUILTIN, HIGH);
    }

    if (blink && millis() - start > 1000) 
    {
        blink = false;
        digitalWrite(LED_BUILTIN, LOW);
    }

    yield();
}
```

This solves Charlie's problem, but it's quite verbose. Using React_Generic, Charlie can write the same script like this:

```c++
#include <React_Generic.h>

using namespace react_generic;

React_Generic app;

void setup() 
{
  Serial.begin(9600);
  pinMode(LED_BUILTIN, OUTPUT);

  app.onAvailable(&Serial, [] () 
  {
    Serial.write(Serial.read());
    digitalWrite(LED_BUILTIN, HIGH);

    app.onDelay(1000, [] () { digitalWrite(LED_BUILTIN, LOW); });
  });
}

void loop() 
{
  app.tick();
}
```

---

### Advanced callback support

Callbacks can be not just void pointers but any callable supported by `std::function`. This means they can use lambda captures or argument binding using `std::bind`. For example, the following code creates 20 different repeating reactions updating different fields of an array:

```c++
static int timer_ticks[20];

for (int i=0; i<20; i++) 
{
  timer_ticks[i] = 0;
  int delay = (i+1)*(i+1);
  app.onRepeat(delay, [i]() 
  {
    timer_ticks[i]++;
  });
}
```

---
---

## API

### Namespace use

Note that beginning of React_Generic `v2.0.0`, the React_Generic library has been wrapped in a `React_Generic` namespace.

This is to avoid name conflicts with other libraries.

The impact to the user is that they need to define the namespace when using the library.
This can be done either globally by placing the following statement in the source code right after the `#include` statements:

```cpp
using namespace react_generic;
```

or by using the `react_generic::` prefix when using the library:

```cpp
react_generic::React_Generic app;
```

### Event Registration Functions

All of the registration functions return a `Reaction` object pointer. This can be used to store and manipulate
the reaction. `react_callback` is a typedef for `std::function<void()>` and can therefore be any callable supported by the C++ standard template library.

```cpp
DelayReaction app.onDelay(uint32_t t, react_callback cb);
```

Delay the execution of a callback by `t` milliseconds.

```cpp
RepeatReaction app.onRepeat(uint32_t t, react_callback cb);
```

Repeatedly execute a callback every `t` milliseconds.

```cpp
StreamReaction app.onAvailable(Stream *stream, react_callback cb);
```

Execute a callback when there is data available to read on an input stream (such as `&Serial`).

```cpp
ISRReaction app.onInterrupt(uint8_t pin_number, int mode, react_callback cb);
```

Execute a callback when an interrupt number fires. This uses the same API as the `attachInterrupt()` Arduino function.

```cpp
TickReaction app.onTick(react_callback cb);
```

Execute a callback on every tick of the event loop.

### Management functions

```cpp
void Reaction::remove();
```

Remove the reaction from the execution queue.

*Note*: Calling `remove()` for `DelayReaction` objects is only safe if the reaction has not been triggered yet. Upon triggering, the `DelayReaction` object is deleted and any pointers to it will be invalidated


---
---

### Examples:

#### Original Examples

 1. [`minimal`](examples/minimal): A minimal example with two timers switching the LED state.
 2. [`torture_test`](examples/torture_test): A stress test of twenty simultaneous repeat reactions as well as a couple of interrupts, a stream, and a tick reaction. For kicks, try changing `NUM_TIMERS` to 200. Program performance will be practically unchanged!

#### multiFileProject Examples

- [`multiFileProject`](examples/multiFileProject)

---
---


### Issues ###

Submit issues to: [React_Generic issues](https://github.com/khoih-prog/React_Generic/issues)

---

### TO DO

1. Bug Searching and Killing
2. Support more types of boards

### DONE

 1. Add support to **Arduino SAMD21 (ZERO, MKR, NANO_33_IOT, etc.)**
 2. Add support to **Adafruit SAMD21 (Itsy-Bitsy M0, Metro M0, Feather M0 Express, etc.)**.
 3. Add support to **Adafruit SAMD51 (Itsy-Bitsy M4, Metro M4, Grand Central M4, Feather M4 Express, etc.)**.
 4. Add support to **Adafruit nRF52 ( Feather nRF52832, nRF52840 Express, BlueFruit Sense, Itsy-Bitsy nRF52840 Express, Metro nRF52840 Express, NINA_B302_ublox, NINA_B112_ublox, etc.**.
 5. Add support to RP2040-based boards such as RASPBERRY_PI_PICO, using [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed)
 6. Add support to RP2040-based boards such as RASPBERRY_PI_PICO, using [**Earle Philhower's arduino-pico** core](https://github.com/earlephilhower/arduino-pico)
 7. Add support to Teensy
 8. Add astyle using `allman` style. Restyle the library

---
---

### Contributions and Thanks

Many thanks for everyone for bug reporting, new feature suggesting, testing and contributing to the development of this library.

1. Based on and modified from [Andrew Carter's Reactduino](https://github.com/Reactduino/Reactduino)
2. Based on and modified from [Matti Airas' ReactESP](https://github.com/mairas/ReactESP)


<table>
  <tr>
    <td align="center"><a href="https://github.com/AndrewCarterUK"><img src="https://github.com/AndrewCarterUK.png" width="100px;" alt="AndrewCarterUK"/><br /><sub><b>⭐️ Andrew Carter</b></sub></a><br /></td>
    <td align="center"><a href="https://github.com/mairas"><img src="https://github.com/mairas.png" width="100px;" alt="mairas"/><br /><sub><b>⭐️ Matti Airas</b></sub></a><br /></td>
  </tr> 
</table>

---

### Contributing

If you want to contribute to this project:
- Report bugs and errors
- Ask for enhancements
- Create issues and pull requests
- Tell other people about this library

---
---

### License

- The library is licensed under [MIT](https://github.com/khoih-prog/React_Generic/blob/main/LICENSE)

---


## Copyright


- Copyright (c) 2017- Andrew Carter
- Copyright (c) 2019- Matti Airas
- Copyright (c) 2022- Khoi Hoang

  


changelog.md
## React_Generic Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/React_Generic.svg?)](https://www.ardu-badge.com/React_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/React_Generic.svg)](https://github.com/khoih-prog/React_Generic/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/React_Generic/blob/master/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/React_Generic.svg)](http://github.com/khoih-prog/React_Generic/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>


---
---

## Table of Contents


* [Changelog](#changelog)
	* [Releases v2.1.0](#releases-v210)


---
---

## Changelog

### Releases v2.1.0

1. Initial porting and coding to support ESP32, ESP8266, SAMD21/SAMD51, RP2040, STM32, nRF52, Teensy 4.x




src/React_Generic.hpp
/****************************************************************************************************************************
  React_Generic.hpp
  
  React_Generic is a library for ESP32, ESP8266, Protenta_H7, STM32F7, etc.
  
  Based on and modified from :
  
  1) Reactduino   (https://github.com/Reactduino/Reactduino)
  2) ReactESP     (https://github.com/mairas/ReactESP)
  
  Built by Khoi Hoang https://github.com/khoih-prog/React_Generic
 
  Version: 2.1.0
  
  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  2.1.0    K Hoang     23/03/2022 Initial porting and coding to support ESP32, ESP8266, RP2040, STM32, nRF52, Teensy 4.x
 *****************************************************************************************************************************/

#pragma once

#ifndef REACT_GENERIC_HPP_
#define REACT_GENERIC_HPP_

#define REACT_GENERIC_VERSION            "React_Generic v2.1.0"

#define REACT_GENERIC_VERSION_MAJOR      2
#define REACT_GENERIC_VERSION_MINOR      1
#define REACT_GENERIC_VERSION_PATCH      0

#define REACT_GENERIC_VERSION_INT        2001000

#include <Arduino.h>

#include <forward_list>
#include <functional>
#include <queue>

#include "React_Generic_Debug.h"

namespace react_generic 
{
  typedef std::function<void()> react_callback;
  typedef void (*isr_react_callback)(void*);
  
  // forward declarations
  
  class React_Generic;
  
  // ESP32 doesn't have the micros64 function defined
  #ifdef ESP32
    uint64_t ICACHE_RAM_ATTR micros64();
  #elif !defined(ESP8266)
    uint64_t micros64();
  #endif
  
  /**
     @brief Reactions are code to be called when a given condition is fulfilled
  */
  class Reaction 
  {
    protected:
      const react_callback callback;
  
    public:
      /**
         @brief Construct a new Reaction object
  
         @param callback Function to be called when the reaction is triggered
      */
      Reaction(react_callback callback) : callback(callback) {}
      // FIXME: why do these have to be defined?
      virtual void add(React_Generic* app = nullptr) = 0;
      virtual void remove(React_Generic* app = nullptr) = 0;
      virtual void tick() = 0;
  };
  
  /**
     @brief TimedReactions are called based on elapsing of time.
  */
  class TimedReaction : public Reaction 
  {
    protected:
      const uint64_t interval;
      uint64_t last_trigger_time;
      bool enabled;
      // A repeat reaction needs to know which app it belongs to
      React_Generic* app_context = nullptr;
  
    public:
      /**
         @brief Construct a new Timed Reaction object
  
         @param interval Interval or delay for the reaction, in milliseconds
         @param callback Function to be called when the reaction is triggered
      */
      TimedReaction(const uint32_t interval, const react_callback callback)
        : Reaction(callback), interval((uint64_t)1000 * (uint64_t)interval) 
      {
        last_trigger_time = micros64();
        enabled = true;
      }
      /**
         @brief Construct a new Timed Reaction object
  
         @param interval Interval, in microseconds
         @param callback Function to be called when the reaction is triggered
      */
      TimedReaction(const uint64_t interval, const react_callback callback)
        : Reaction(callback), interval(interval) 
      {
        last_trigger_time = micros64();
        enabled = true;
      }
  
      virtual ~TimedReaction() {}
      
      bool operator<(const TimedReaction& other);
      void add(React_Generic* app = nullptr) override;
      void remove(React_Generic* app = nullptr) override;
      
      uint32_t getTriggerTime() 
      {
        return (last_trigger_time + interval) / 1000;
      }
      
      uint64_t getTriggerTimeMicros() 
      {
        return (last_trigger_time + interval);
      }
      
      bool isEnabled() 
      {
        return enabled;
      }
      
      virtual void tick() = 0;
  };
  
  struct TriggerTimeCompare 
  {
    bool operator()(TimedReaction* a, TimedReaction* b) 
    {
      return *b < *a;
    }
  };
  
  /**
     @brief Reaction that is triggered after a certain time delay
  */
  class DelayReaction : public TimedReaction 
  {
    public:
      /**
         @brief Construct a new Delay Reaction object
  
         @param delay Delay, in milliseconds
         @param callback Function to be called after the delay
      */
      DelayReaction(const uint32_t delay, const react_callback callback);
      /**
         @brief Construct a new Delay Reaction object
  
         @param delay Delay, in microseconds
         @param callback Function to be called after the delay
      */
      DelayReaction(const uint64_t delay, const react_callback callback);
      
      virtual ~DelayReaction() {}
      
      void tick();
  };
  
  /**
     @brief Reaction that is triggered repeatedly
  */
  class RepeatReaction : public TimedReaction 
  {
    public:
      /**
         @brief Construct a new Repeat Reaction object
  
         @param interval Repetition interval, in milliseconds
         @param callback Function to be called at every repetition
      */
      RepeatReaction(const uint32_t interval, const react_callback callback)
        : TimedReaction(interval, callback) {}
      /**
         @brief Construct a new Repeat Reaction object
  
         @param interval Repetition interval, in microseconds
         @param callback Function to be called at every repetition
      */
      RepeatReaction(const uint64_t interval, const react_callback callback)
        : TimedReaction(interval, callback) {}
        
      void tick();
  };
  
  /**
     @brief Reactions that are triggered based on something else than time
  */
  class UntimedReaction : public Reaction 
  {
    public:
      UntimedReaction(const react_callback callback) : Reaction(callback) {}
      virtual ~UntimedReaction() {}
      
      virtual void add(React_Generic* app = nullptr) override;
      virtual void remove(React_Generic* app = nullptr) override;
      virtual void tick() = 0;
  };
  
  /**
     @brief Reaction that is triggered when there is input available at the given
       Arduino Stream
  */
  class StreamReaction : public UntimedReaction 
  {
    private:
      Stream& stream;
  
    public:
      /**
         @brief Construct a new Stream Reaction object
  
         @param stream Stream to monitor
         @param callback Callback to call for new input
      */
      StreamReaction(Stream& stream, const react_callback callback)
        : UntimedReaction(callback), stream(stream) {}
      void tick();
  };
  
  /**
     @brief Reaction that is triggered unconditionally at each execution loop
  */
  class TickReaction : public UntimedReaction 
  {
    public:
      /**
         @brief Construct a new Tick Reaction object
  
         @param callback Function to be called at each execution loop
      */
      TickReaction(const react_callback callback) : UntimedReaction(callback) {}
      
      void tick();
  };
  
  /**
     @brief Reaction that is triggered on an input pin change
  */
  class ISRReaction : public Reaction 
  {
    private:
      const uint8_t pin_number;
      const int mode;
  #ifdef ESP32
      // set to true once gpio_install_isr_service is called
      static bool isr_service_installed;
      static void isr(void* arg);
  #endif
  
    public:
      /**
         @brief Construct a new ISRReaction object
  
         @param pin_number GPIO pin number to which the interrupt is attached
         @param mode Interrupt mode. One of RISING, FALLING, CHANGE
         @param callback Interrupt callback. Keep this function short and add the
         ICACHE_RAM_ATTR attribute.
      */
      ISRReaction(uint8_t pin_number, int mode, const react_callback callback)
        : Reaction(callback), pin_number(pin_number), mode(mode) 
      {
  #ifdef ESP32
        gpio_int_type_t intr_type;
        
        switch (mode) 
        {
          case RISING:
            intr_type = GPIO_INTR_POSEDGE;
            break;
          case FALLING:
            intr_type = GPIO_INTR_NEGEDGE;
            break;
          case CHANGE:
            intr_type = GPIO_INTR_ANYEDGE;
            break;
          default:
            intr_type = GPIO_INTR_DISABLE;
            break;
        }
        
        // configure the IO pin
        gpio_set_intr_type((gpio_num_t)pin_number, intr_type);
  
        if (!isr_service_installed) 
        {
          gpio_install_isr_service(ESP_INTR_FLAG_LOWMED);
          isr_service_installed = true;
        }
  #endif
      }
      
      virtual ~ISRReaction() {}
      void add(React_Generic* app = nullptr) override;
      void remove(React_Generic* app = nullptr) override;
      void tick() {}
  };
  
  ///////////////////////////////////////
  // React_Generic main class implementation
  
  /**
     @brief Main class of a React_Generic program
  */
  class React_Generic 
  {
      friend class Reaction;
      friend class TimedReaction;
      friend class RepeatReaction;
      friend class UntimedReaction;
      friend class ISRReaction;
  
    public:
      /**
         @brief Construct a new React_Generic object.
  
         @param singleton If true, set the singleton instance to this object
      */
      React_Generic(bool singleton = true) 
      {
        if (singleton) 
          app = this;
      }
      
      void tick(void);
  
      /// Static singleton reference to the instantiated React_Generic object
      static React_Generic* app;
  
      /**
         @brief Create a new DelayReaction
  
         @param t Delay, in milliseconds
         @param cb Callback function
         @return DelayReaction
      */
      DelayReaction* onDelay(const uint32_t t, const react_callback cb);
      /**
         @brief Create a new DelayReaction
  
         @param t Delay, in microseconds
         @param cb Callback function
         @return DelayReaction
      */
      DelayReaction* onDelayMicros(const uint64_t t, const react_callback cb);
      /**
         @brief Create a new RepeatReaction
  
         @param t Interval, in milliseconds
         @param cb Callback function
         @return RepeatReaction
      */
      RepeatReaction* onRepeat(const uint32_t t, const react_callback cb);
      /**
         @brief Create a new RepeatReaction
  
         @param t Interval, in microseconds
         @param cb Callback function
         @return RepeatReaction
      */
      RepeatReaction* onRepeatMicros(const uint64_t t, const react_callback cb);
      /**
         @brief Create a new StreamReaction
  
         @param stream Arduino Stream object to monitor
         @param cb Callback function
         @return StreamReaction
      */
      StreamReaction* onAvailable(Stream& stream, const react_callback cb);
      /**
         @brief Create a new ISRReaction (interrupt reaction)
  
         @param pin_number GPIO pin number
         @param mode One of CHANGE, RISING, FALLING
         @param cb Interrupt handler to call. This should be a very simple function,
         ideally setting a flag variable or incrementing a counter. The function
         should be defined with ICACHE_RAM_ATTR.
         @return ISRReaction
      */
      ISRReaction* onInterrupt(const uint8_t pin_number, int mode,
                               const react_callback cb);
      /**
         @brief Create a new TickReaction
  
         @param cb Callback function to be called at every loop execution
         @return TickReaction
      */
      TickReaction* onTick(const react_callback cb);
  
      /**
         @brief Remove a reaction from the list of active reactions
  
         @param reaction Reaction to remove
      */
      void remove(Reaction* reaction);
  
    private:
    
      std::priority_queue<TimedReaction*, std::vector<TimedReaction*>, TriggerTimeCompare> timed_queue;
      std::forward_list<UntimedReaction*> untimed_list;
      std::forward_list<ISRReaction*> isr_reaction_list;
      std::forward_list<ISRReaction*> isr_pending_list;
      
      void tickTimed();
      void tickUntimed();
      void tickISR();
      void add(Reaction* re);
  };

}  // namespace react_generic

#endif		// REACT_GENERIC_HPP_


examples/multiFileProject/multiFileProject.h
/****************************************************************************************************************************
    multiFileProject.h
    React_Generic is a library for ESP32, ESP8266, Protenta_H7, STM32F7, etc.

    Based on and modified from :

    1) Reactduino   (https://github.com/Reactduino/Reactduino)
    2) ReactESP     (https://github.com/mairas/ReactESP)

    Built by Khoi Hoang https://github.com/khoih-prog/React_Generic
*****************************************************************************************************************************/

// To demo how to include files in multi-file Projects

#pragma once

#define _TEENSY41_ASYNC_TCP_LOGLEVEL_       1

// Can be included as many times as necessary, without `Multiple Definitions` Linker Error
#include <React_Generic.hpp>        // https://github.com/khoih-prog/React_Generic


src/React_Generic.h
/****************************************************************************************************************************
    React_Generic.h

    React_Generic is a library for ESP32, ESP8266, Protenta_H7, STM32F7, etc.

    Based on and modified from :

    1) Reactduino   (https://github.com/Reactduino/Reactduino)
    2) ReactESP     (https://github.com/mairas/ReactESP)

    Built by Khoi Hoang https://github.com/khoih-prog/React_Generic

    Version: 2.1.0

    Version Modified By   Date      Comments
    ------- -----------  ---------- -----------
    2.1.0    K Hoang     23/03/2022 Initial porting and coding to support ESP32, ESP8266, RP2040, STM32, nRF52, Teensy 4.x
 *****************************************************************************************************************************/

#pragma once

#ifndef REACT_GENERIC_H_
#define REACT_GENERIC_H_

#include <Arduino.h>

#if (defined(NRF52840_FEATHER) || defined(NRF52832_FEATHER) || defined(NRF52_SERIES) || defined(ARDUINO_NRF52_ADAFRUIT) || \
      defined(NRF52840_FEATHER_SENSE) || defined(NRF52840_ITSYBITSY) || defined(NRF52840_CIRCUITPLAY) || \
      defined(NRF52840_CLUE) || defined(NRF52840_METRO) || defined(NRF52840_PCA10056) || defined(PARTICLE_XENON) || \
      defined(MDBT50Q_RX) || defined(NINA_B302_ublox) || defined(NINA_B112_ublox) )
#include <Adafruit_TinyUSB.h>

#define REACT_GENERIC_USING_NRF52             true

#elif ( defined(__arm__) && defined(TEENSYDUINO) && defined(__IMXRT1062__) )

// For Teensy 4.0/4.1
#if defined(ARDUINO_TEENSY41)
  #warning Using Teensy 4.1

  #ifndef BOARD_NAME
    #define BOARD_NAME                        "Teensy 4.1"
  #endif
#else
  #warning Using Teensy 4.0

  #ifndef BOARD_NAME
    #define BOARD_NAME                        "Teensy 4.0"
  #endif
#endif

#define REACT_GENERIC_USING_TEENSY4           true
#endif

#include "React_Generic.hpp"
#include "React_Generic_Impl.h"

#endif      // REACT_GENERIC_H_
